packages/langium/src/generator/node-joiner.ts

```diff
@@ -50,25 +90,39 @@ export interface JoinOptions<T> {
  */
 export function joinToNode<T>(
     iterable: Iterable<T> | T[],
-    toGenerated: (element: T, index: number, isLast: boolean) => Generated = String,
-    { filter, prefix, suffix, separator, appendNewLineIfNotEmpty }: JoinOptions<T> = {}
+    toGenerated?: ((element: T, index: number, isLast: boolean) => Generated),
+    options?: JoinOptions<T>
+): CompositeGeneratorNode | undefined;
+
+export function joinToNode<T>(
+    iterable: Iterable<T> | T[],
+    toGeneratedOrOptions: ((element: T, index: number, isLast: boolean) => Generated) | JoinOptions<T> = defaultToGenerated,
+    options: JoinOptions<T> = {}
 ): CompositeGeneratorNode | undefined {
 
+    const toGenerated = typeof toGeneratedOrOptions === 'function' ? toGeneratedOrOptions : defaultToGenerated;
+    const { filter, prefix, suffix, separator, appendNewLineIfNotEmpty, skipNewLineAfterLastItem } = typeof toGeneratedOrOptions === 'object' ? toGeneratedOrOptions : options;
+
+    const prefixFunc = typeof prefix === 'function' ? prefix : undefined;
+    const suffixFunc = typeof suffix === 'function' ? suffix : undefined;
```

Comment:
Suggestion: All the undefined-checking and casting to `Generated` below is a bit ugly, we can just define the function here a bit smarter.

```suggestion
    const prefixFunc = typeof prefix === 'function' ? prefix : (() => prefix);
    const suffixFunc = typeof suffix === 'function' ? suffix : (() => suffix);
```